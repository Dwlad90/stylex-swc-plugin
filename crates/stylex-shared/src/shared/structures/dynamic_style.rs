use swc_core::ecma::ast::Expr;

#[derive(Debug, Clone)]
pub(crate) struct DynamicStyle {
  pub(crate) expression: Expr,
  pub(crate) key: String,
  pub(crate) path: String,
}
