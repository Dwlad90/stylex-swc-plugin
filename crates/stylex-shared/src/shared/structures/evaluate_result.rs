use std::collections::HashMap;

use indexmap::IndexMap;
use swc_core::{
  atoms::Atom,
  ecma::ast::{Expr, KeyValueProp, Lit},
};

use crate::shared::enums::data_structures::evaluate_result_value::EvaluateResultValue;

use super::{
  functions::FunctionConfig,
  theme_ref::ThemeRef,
  types::{DynamicFns, EvaluationCallback, TInlineStyles},
};

#[derive(Debug, Clone, PartialEq)]
pub struct EvaluateResult {
  pub(crate) confident: bool,
  pub value: Option<Box<EvaluateResultValue>>,
  pub(crate) deopt: Option<Box<Expr>>,
  pub(crate) inline_styles: Option<TInlineStyles>,
  pub(crate) fns: Option<DynamicFns>,
}

impl EvaluateResultValue {
  pub fn as_expr(&self) -> Option<&Expr> {
    match self {
      EvaluateResultValue::Expr(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_vec(&self) -> Option<&Vec<Option<EvaluateResultValue>>> {
    match self {
      EvaluateResultValue::Vec(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_map(&self) -> Option<&IndexMap<Box<Expr>, Vec<KeyValueProp>>> {
    match self {
      EvaluateResultValue::Map(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_entries(&self) -> Option<&IndexMap<Box<Lit>, Box<Lit>>> {
    match self {
      EvaluateResultValue::Entries(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_function(&self) -> Option<&FunctionConfig> {
    match self {
      EvaluateResultValue::FunctionConfig(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_function_map(&self) -> Option<&HashMap<Atom, FunctionConfig>> {
    match self {
      EvaluateResultValue::FunctionConfigMap(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_callback(&self) -> Option<&EvaluationCallback> {
    match self {
      EvaluateResultValue::Callback(value) => Some(value),
      _ => None,
    }
  }

  pub fn as_theme_ref(&self) -> Option<&ThemeRef> {
    match self {
      EvaluateResultValue::ThemeRef(value) => Some(value),
      _ => None,
    }
  }
}
