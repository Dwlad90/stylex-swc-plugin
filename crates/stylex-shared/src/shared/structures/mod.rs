pub(crate) mod application_order;
pub(crate) mod base_css_type;
pub(crate) mod dynamic_style;
pub mod evaluate_result;
pub mod functions;
pub(crate) mod injectable_style;
pub(crate) mod inline_style;
pub(crate) mod legacy_expand_shorthands_order;
pub(crate) mod member_transform;
pub mod meta_data;
pub mod named_import_source;
pub(crate) mod null_pre_rule;
pub(crate) mod order;
pub(crate) mod order_pair;
pub(crate) mod pair;
pub mod plugin_pass;
pub(crate) mod pre_rule;
pub(crate) mod pre_rule_set;
pub(crate) mod property_specificity;
pub(crate) mod property_specificity_order;
pub(crate) mod seen_value;
pub(crate) mod shorthands_of_shorthands;
pub(crate) mod state;
pub mod state_manager;
pub mod stylex_options;
pub(crate) mod stylex_state_options;
pub(crate) mod tests;
pub(crate) mod theme_ref;
pub(crate) mod types;
pub(crate) mod uid_generator;
