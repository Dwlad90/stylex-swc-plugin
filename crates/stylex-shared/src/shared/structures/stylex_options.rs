use std::collections::HashMap;

use serde::Deserialize;

use crate::shared::constants::common::DEFAULT_INJECT_PATH;

use super::named_import_source::{ImportSources, RuntimeInjection};

#[derive(Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct StyleXOptionsParams {
  pub style_resolution: Option<StyleResolution>,
  pub use_rem_for_font_size: Option<bool>,
  pub runtime_injection: Option<bool>,
  pub class_name_prefix: Option<String>,
  pub defined_stylex_css_variables: Option<HashMap<String, String>>,
  pub import_sources: Option<Vec<ImportSources>>,
  pub treeshake_compensation: Option<bool>,
  pub gen_conditional_classes: Option<bool>,
  pub dev: Option<bool>,
  pub test: Option<bool>,
  pub aliases: Option<HashMap<String, Vec<String>>>,
  #[serde(rename = "unstable_moduleResolution")]
  pub unstable_module_resolution: Option<ModuleResolution>,
}

impl Default for StyleXOptionsParams {
  fn default() -> Self {
    StyleXOptionsParams {
      style_resolution: Some(StyleResolution::ApplicationOrder),
      use_rem_for_font_size: Some(false),
      runtime_injection: Some(false),
      class_name_prefix: Some("x".to_string()),
      defined_stylex_css_variables: Some(HashMap::new()),
      import_sources: None,
      treeshake_compensation: Some(true),
      gen_conditional_classes: Some(false),
      dev: Some(false),
      test: Some(false),
      aliases: None,
      unstable_module_resolution: None,
    }
  }
}

#[derive(Deserialize, Debug, Clone)]
#[serde(rename_all(deserialize = "kebab-case", serialize = "PascalCase"))]

pub enum StyleResolution {
  ApplicationOrder,
  PropertySpecificity,
  LegacyExpandShorthands,
}

#[derive(Deserialize, Debug, Clone)]

pub enum Aliases {
  String(HashMap<String, String>),
  StringVec(HashMap<String, Vec<String>>),
}

#[derive(Deserialize, Debug, Clone)]
#[serde(rename_all(deserialize = "camelCase", serialize = "PascalCase"))]
pub struct ModuleResolution {
  pub r#type: String,
  pub root_dir: Option<String>,
  pub theme_file_extension: Option<String>,
}

#[derive(Deserialize, Debug, Clone)]

pub enum CheckModuleResolution {
  CommonJS(ModuleResolution),
  Haste(ModuleResolution),
  CrossFileParsing(ModuleResolution),
}

#[derive(Clone, Debug)]
pub struct StyleXOptions {
  pub dev: bool,
  pub test: bool,
  pub use_rem_for_font_size: bool,
  pub class_name_prefix: String,
  // pub defined_stylex_css_variables: HashMap<String, String>,
  pub style_resolution: StyleResolution,
  pub runtime_injection: RuntimeInjection,
  pub import_sources: Vec<ImportSources>,
  pub treeshake_compensation: Option<bool>,
  pub gen_conditional_classes: bool,
  pub aliases: Option<HashMap<String, Vec<String>>>,
  pub unstable_module_resolution: Option<CheckModuleResolution>,
}

impl StyleXOptions {
  pub fn get_haste_module_resolution(root_dir: Option<String>) -> ModuleResolution {
    ModuleResolution {
      r#type: "haste".to_string(),
      root_dir,
      theme_file_extension: None,
    }
  }

  pub fn get_common_js_module_resolution(root_dir: Option<String>) -> ModuleResolution {
    ModuleResolution {
      r#type: "commonjs".to_string(),
      root_dir,
      theme_file_extension: None,
    }
  }
}

impl Default for StyleXOptions {
  fn default() -> Self {
    StyleXOptions {
      style_resolution: StyleResolution::ApplicationOrder,
      use_rem_for_font_size: false,
      runtime_injection: RuntimeInjection::Boolean(false),
      class_name_prefix: "x".to_string(),
      // defined_stylex_css_variables: HashMap::new(),
      import_sources: vec![],
      dev: false,
      test: false,
      treeshake_compensation: None,
      gen_conditional_classes: false,
      aliases: None,
      unstable_module_resolution: Some(CheckModuleResolution::Haste(
        StyleXOptions::get_haste_module_resolution(None),
      )),
    }
  }
}

impl From<StyleXOptionsParams> for StyleXOptions {
  fn from(options: StyleXOptionsParams) -> Self {
    let unstable_module_resolution = match options.unstable_module_resolution {
      Some(module_resolution) => match module_resolution.r#type.to_lowercase().as_str() {
        "haste" => Some(CheckModuleResolution::Haste(module_resolution)),
        "cross-file-parsing" => Some(CheckModuleResolution::CrossFileParsing(module_resolution)),
        _ => Some(CheckModuleResolution::CommonJS(module_resolution)),
      },
      None => None,
    };

    let runtime_injection = match options.runtime_injection {
      Some(value) => match value {
        true => RuntimeInjection::Regular(DEFAULT_INJECT_PATH.to_string()),
        false => RuntimeInjection::Boolean(options.dev.unwrap_or(false)),
      },
      None => RuntimeInjection::Boolean(options.dev.unwrap_or(false)),
    };

    StyleXOptions {
      style_resolution: options
        .style_resolution
        .unwrap_or(StyleResolution::ApplicationOrder),
      use_rem_for_font_size: options.use_rem_for_font_size.unwrap_or(false),
      runtime_injection,
      class_name_prefix: options.class_name_prefix.unwrap_or("x".to_string()),
      // defined_stylex_css_variables: options.defined_stylex_css_variables.unwrap_or_default(),
      import_sources: options.import_sources.unwrap_or_default(),
      dev: options.dev.unwrap_or(false),
      test: options.test.unwrap_or(false),
      treeshake_compensation: options.treeshake_compensation,
      gen_conditional_classes: options.gen_conditional_classes.unwrap_or(false),
      aliases: options.aliases,
      unstable_module_resolution,
    }
  }
}
