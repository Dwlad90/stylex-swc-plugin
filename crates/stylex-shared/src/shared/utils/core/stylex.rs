use crate::{
  shared::{
    enums::data_structures::fn_result::FnResult, utils::ast::convertors::string_to_expression,
  },
  transform::styleq::common::styleq,
};

use super::parse_nullable_style::ResolvedArg;

pub(crate) fn stylex(values: &Vec<ResolvedArg>) -> Option<FnResult> {
  let result = styleq(values);

  Some(FnResult::Stylex(string_to_expression(
    result.class_name.as_str(),
  )))
}
