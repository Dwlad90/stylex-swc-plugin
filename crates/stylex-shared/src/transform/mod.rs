use rustc_hash::FxHashSet;
use swc_core::{
  common::{Mark, comments::Comments},
  ecma::{
    ast::{CallExpr, Callee, Expr, Id, MemberProp, Pass, VarDeclarator},
    transforms::base::resolver,
    utils::drop_span,
    visit::fold_pass,
  },
};

use crate::{
  StyleXOptionsParams,
  shared::{
    enums::core::TransformationCycle,
    structures::{
      named_import_source::{ImportSources, RuntimeInjection},
      plugin_pass::PluginPass,
      state_manager::StateManager,
      stylex_options::{CheckModuleResolution, StyleXOptions},
    },
    utils::common::increase_ident_count,
  },
};

mod fold;
pub(crate) mod styleq;
pub(crate) mod stylex;

pub struct StyleXTransform<C>
where
  C: Comments,
{
  pub comments: C,
  props_declaration: Option<Id>,
  pub state: StateManager,
}

impl<C> StyleXTransform<C>
where
  C: Comments,
{
  pub fn new(comments: C, plugin_pass: PluginPass, config: &mut StyleXOptionsParams) -> Self {
    let stylex_imports = fill_stylex_imports(&Some(config));

    let mut state = StateManager::new(config.clone().into());

    state.stylex_import.clone_from(&stylex_imports);

    state.options.import_sources = stylex_imports.into_iter().collect();

    state._state = plugin_pass;

    StyleXTransform {
      comments,
      props_declaration: None,
      state,
    }
  }

  pub fn new_test_force_runtime_injection(
    comments: C,
    plugin_pass: PluginPass,
    config: Option<&mut StyleXOptionsParams>,
  ) -> Self {
    let stylex_imports = fill_stylex_imports(&config);

    let mut state = match config {
      Some(config) => {
        config.runtime_injection = Some(true);
        config.treeshake_compensation = Some(true);

        StateManager::new(config.clone().into())
      }
      None => {
        let config = StyleXOptions {
          runtime_injection: RuntimeInjection::Boolean(true),
          treeshake_compensation: Some(true),
          unstable_module_resolution: CheckModuleResolution::Haste(
            StyleXOptions::get_haste_module_resolution(None),
          ),
          ..Default::default()
        };

        StateManager::new(config)
      }
    };

    state.options.import_sources = stylex_imports.into_iter().collect();

    state._state = plugin_pass;

    StyleXTransform {
      comments,
      props_declaration: None,
      state,
    }
  }

  pub fn new_test_force_runtime_injection_with_pass(
    comments: C,
    plugin_pass: PluginPass,
    config: Option<&mut StyleXOptionsParams>,
  ) -> impl Pass + use<C> {
    (
      resolve_factory(),
      fold_pass(Self::new_test_force_runtime_injection(
        comments,
        plugin_pass,
        config,
      )),
    )
  }

  pub fn new_test(
    comments: C,
    plugin_pass: PluginPass,
    config: Option<&mut StyleXOptionsParams>,
  ) -> Self {
    let stylex_imports = fill_stylex_imports(&config);

    let mut state = match config {
      Some(config) => StateManager::new(config.clone().into()),
      None => {
        let config = StyleXOptions {
          runtime_injection: RuntimeInjection::Boolean(false),
          treeshake_compensation: Some(true),
          class_name_prefix: "x".to_string(),
          unstable_module_resolution: CheckModuleResolution::Haste(
            StyleXOptions::get_haste_module_resolution(None),
          ),
          ..Default::default()
        };

        StateManager::new(config)
      }
    };

    state.options.import_sources = stylex_imports.into_iter().collect();

    state._state = plugin_pass;

    StyleXTransform {
      comments,
      props_declaration: None,
      state,
    }
  }
  pub fn new_test_with_pass(
    comments: C,
    plugin_pass: PluginPass,
    config: Option<&mut StyleXOptionsParams>,
  ) -> impl Pass + use<C> {
    (
      resolve_factory(),
      fold_pass(Self::new_test(comments, plugin_pass, config)),
    )
  }

  pub(crate) fn process_declaration(&mut self, call_expr: &mut CallExpr) -> Option<(Id, String)> {
    let state = &self.state;

    let stylex_imports = state.stylex_import_stringified();
    if let Callee::Expr(callee) = &mut call_expr.callee {
      match callee.as_ref() {
        Expr::Ident(ident) => {
          let ident_id = ident.to_id();

          if stylex_imports.contains(&ident.sym.to_string())
            || (state.cycle == TransformationCycle::TransformEnter
              && (state.stylex_create_import.contains(&ident.sym))
              || state.stylex_props_import.contains(&ident.sym)
              || state.stylex_keyframes_import.contains(&ident.sym)
              || state.stylex_first_that_works_import.contains(&ident.sym)
              || state.stylex_types_import.contains(&ident.sym)
              || state.stylex_create_theme_import.contains(&ident.sym)
              || state.stylex_define_vars_import.contains(&ident.sym)
              || state.stylex_attrs_import.contains(&ident.sym))
          {
            increase_ident_count(&mut self.state, ident);

            return Some((ident_id.clone(), format!("{}", ident.sym)));
          }
        }
        Expr::Member(member) => {
          if let Expr::Ident(ident) = member.obj.as_ref() {
            let ident_id = ident.to_id();

            if stylex_imports.contains(&ident.sym.to_string())
              || (state.cycle == TransformationCycle::TransformEnter
                && (state.stylex_create_import.contains(&ident.sym))
                || state.stylex_props_import.contains(&ident.sym)
                || state.stylex_keyframes_import.contains(&ident.sym)
                || state.stylex_first_that_works_import.contains(&ident.sym)
                || state.stylex_create_theme_import.contains(&ident.sym)
                || state.stylex_types_import.contains(&ident.sym)
                || state.stylex_define_vars_import.contains(&ident.sym)
                || state.stylex_attrs_import.contains(&ident.sym))
            {
              if let MemberProp::Ident(ident) = &member.prop {
                return Some((ident_id.clone(), format!("{}", ident.sym)));
              }
            }
          }
        }
        _ => {}
      }
    }

    None
  }

  pub(crate) fn transform_call_expression(&mut self, expr: &mut Expr) -> Option<Expr> {
    if let Expr::Call(ex) = expr {
      let declaration = self.process_declaration(ex);

      if declaration.is_some() {
        return self.transform_call_expression_to_stylex_expr(ex);
      }
    }

    None
  }

  pub(crate) fn get_call_var_name(
    &mut self,
    call: &CallExpr,
  ) -> (Option<String>, Option<VarDeclarator>) {
    let mut var_name: Option<String> = None;

    let parent_var_decl = self
      .state
      .declarations
      .iter()
      .find(|decl| match &decl.init {
        Some(init) => {
          if let Expr::Call(init_call) = init.as_ref() {
            init_call == &drop_span(call.clone())
          } else {
            false
          }
        }
        _ => false,
      })
      .cloned();

    if let Some(ref parent_var_decl) = parent_var_decl {
      if let Some(ident) = parent_var_decl.name.as_ident() {
        var_name = Some(ident.sym.to_string());
      }
    }

    (var_name, parent_var_decl)
  }
}

fn fill_stylex_imports(config: &Option<&mut StyleXOptionsParams>) -> FxHashSet<ImportSources> {
  let mut stylex_imports = FxHashSet::default();

  stylex_imports.insert(ImportSources::Regular("stylex".to_string()));
  stylex_imports.insert(ImportSources::Regular("@stylexjs/stylex".to_string()));

  if let Some(stylex_imports_extends) = match config {
    Some(config) => config.import_sources.clone(),
    None => None,
  } {
    stylex_imports.extend(stylex_imports_extends)
  }

  stylex_imports
}

#[warn(clippy::extra_unused_type_parameters)]
fn resolve_factory() -> impl Pass {
  let unresolved_mark = Mark::new();
  let top_level_mark = Mark::new();

  resolver(unresolved_mark, top_level_mark, true)
}
