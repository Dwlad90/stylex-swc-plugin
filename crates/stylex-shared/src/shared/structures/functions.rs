use std::{fmt::Debug, rc::Rc, sync::Arc};

use indexmap::IndexMap;
use rustc_hash::FxHashMap;
use swc_core::{atoms::Atom, ecma::ast::Expr};

use crate::shared::{
  enums::{
    data_structures::value_with_default::ValueWithDefault,
    js::{ArrayJS, MathJS, ObjectJS, StringJS},
  },
  structures::types::FlatCompiledStyles,
};

use super::{
  state_manager::StateManager,
  types::{FunctionMapIdentifiers, FunctionMapMemberExpression},
};

#[derive(Debug, Hash, PartialEq, Clone)]
pub enum CallbackType {
  Array(ArrayJS),
  Object(ObjectJS),
  Math(MathJS),
  String(StringJS),
  Custom(Expr),
}

pub type StylexTypeFn = Rc<dyn Fn(ValueWithDefault) -> Expr + 'static>;
pub type StylexExprFn = fn(Expr, &mut StateManager) -> Expr;

pub enum FunctionType {
  ArrayArgs(fn(Vec<Expr>) -> Expr),
  StylexExprFn(StylexExprFn),
  StylexTypeFn(StylexTypeFn),
  StylexFnsFactory(fn(input: String) -> StylexTypeFn),

  Mapper(Rc<dyn Fn() -> Expr + 'static>),
  Callback(Box<CallbackType>),
  DefaultMarker(Arc<IndexMap<String, StylexExprFn>>),
}

impl Clone for FunctionType {
  fn clone(&self) -> Self {
    match self {
      Self::ArrayArgs(e) => Self::ArrayArgs(*e),
      Self::StylexExprFn(e) => Self::StylexExprFn(*e),
      Self::StylexTypeFn(e) => Self::StylexTypeFn(e.clone()),
      Self::StylexFnsFactory(e) => Self::StylexFnsFactory(*e),
      Self::Callback(v) => Self::Callback(v.clone()),
      Self::Mapper(c) => Self::Mapper(Rc::clone(c)),
      Self::DefaultMarker(e) => Self::DefaultMarker(Arc::clone(e)),
    }
  }
}

impl std::fmt::Debug for FunctionType {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    match self {
      FunctionType::ArrayArgs(_) => write!(f, "ArrayArgs"),
      FunctionType::StylexExprFn(_) => write!(f, "StylexExprWithStateFn"),
      FunctionType::StylexTypeFn(_) => write!(f, "StylexExprFn"),
      FunctionType::StylexFnsFactory(_) => write!(f, "StylexFnsFactory"),
      FunctionType::Mapper(_) => write!(f, "Mapper"),
      FunctionType::Callback(_) => write!(f, "Callback"),
      FunctionType::DefaultMarker(_) => write!(f, "DefaultMarker"),
    }
  }
}

impl PartialEq for FunctionType {
  fn eq(&self, other: &Self) -> bool {
    match (self, other) {
      (FunctionType::ArrayArgs(_), FunctionType::ArrayArgs(_)) => false,
      (FunctionType::StylexExprFn(_), FunctionType::StylexExprFn(_)) => false,
      (FunctionType::StylexTypeFn(_), FunctionType::StylexTypeFn(_)) => false,
      (FunctionType::StylexFnsFactory(_), FunctionType::StylexFnsFactory(_)) => false,
      (FunctionType::Mapper(_), FunctionType::StylexExprFn(_)) => false,
      (FunctionType::Callback(_), FunctionType::Callback(_)) => false,
      (FunctionType::DefaultMarker(_), FunctionType::DefaultMarker(_)) => false,
      _ => false,
    }
  }
}

impl std::hash::Hash for FunctionType {
  fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
    match self {
      FunctionType::ArrayArgs(_) => {
        std::mem::discriminant(self).hash(state);
      }
      FunctionType::StylexExprFn(_) => {
        std::mem::discriminant(self).hash(state);
      }
      FunctionType::StylexTypeFn(_) => {
        std::mem::discriminant(self).hash(state);
      }
      FunctionType::StylexFnsFactory(_) => {
        std::mem::discriminant(self).hash(state);
      }
      FunctionType::Mapper(_) => {
        std::mem::discriminant(self).hash(state);
      }
      FunctionType::Callback(_) => {
        std::mem::discriminant(self).hash(state);
      }
      FunctionType::DefaultMarker(_) => {
        std::mem::discriminant(self).hash(state);
      }
    }
  }
}

#[derive(Debug, Hash, PartialEq, Clone)]
pub struct FunctionConfig {
  pub fn_ptr: FunctionType,
  pub takes_path: bool,
}

pub enum FunctionConfigType {
  Regular(FunctionConfig),
  Map(FxHashMap<Atom, FunctionConfig>),
  IndexMap(FlatCompiledStyles),
}

impl std::fmt::Debug for FunctionConfigType {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    match self {
      Self::Regular(config) => f.debug_tuple("Regular").field(config).finish(),
      Self::Map(map) => f.debug_tuple("Map").field(map).finish(),
      Self::IndexMap(map) => f.debug_tuple("IndexMap").field(map).finish(),
    }
  }
}

impl Clone for FunctionConfigType {
  fn clone(&self) -> Self {
    match self {
      Self::Regular(config) => Self::Regular(config.clone()),
      Self::Map(map) => Self::Map(map.clone()),
      Self::IndexMap(map) => Self::IndexMap(map.clone()),
    }
  }
}

impl PartialEq for FunctionConfigType {
  fn eq(&self, other: &Self) -> bool {
    match (self, other) {
      (Self::Regular(a), Self::Regular(b)) => a == b,
      (Self::Map(a), Self::Map(b)) => a == b,
      (Self::IndexMap(a), Self::IndexMap(b)) => a == b,
      _ => false,
    }
  }
}

impl FunctionConfigType {
  pub(crate) fn _as_function_config(&self) -> Option<&FunctionConfig> {
    match self {
      Self::Regular(config) => Some(config),
      Self::Map(_) => None,
      Self::IndexMap(_) => None,
    }
  }

  pub(crate) fn _as_map(&self) -> Option<&FxHashMap<Atom, FunctionConfig>> {
    match self {
      Self::Regular(_) => None,
      Self::Map(map) => Some(map),
      Self::IndexMap(_) => None,
    }
  }

  pub(crate) fn as_map_mut(&mut self) -> Option<&mut FxHashMap<Atom, FunctionConfig>> {
    match self {
      Self::Regular(_) => None,
      Self::Map(map) => Some(map),
      Self::IndexMap(_) => None,
    }
  }

  pub(crate) fn _as_function_config_mut(&mut self) -> Option<&mut FunctionConfig> {
    match self {
      Self::Regular(config) => Some(config),
      Self::Map(_) => None,
      Self::IndexMap(_) => None,
    }
  }

  pub(crate) fn _as_index_map(&self) -> Option<&FlatCompiledStyles> {
    match self {
      Self::Regular(_) => None,
      Self::Map(_) => None,
      Self::IndexMap(map) => Some(map),
    }
  }

  pub(crate) fn _as_index_map_mut(&mut self) -> Option<&mut FlatCompiledStyles> {
    match self {
      Self::Regular(_) => None,
      Self::Map(_) => None,
      Self::IndexMap(map) => Some(map),
    }
  }
}

#[derive(Debug, PartialEq, Clone, Default)]
pub struct FunctionMap {
  pub identifiers: FunctionMapIdentifiers,
  pub member_expressions: FunctionMapMemberExpression,
}
