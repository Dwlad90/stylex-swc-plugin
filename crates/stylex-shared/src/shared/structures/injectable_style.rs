use std::{hash::Hash, rc::Rc};

use serde::{Deserialize, Serialize};

use crate::shared::{enums::data_structures::injectable_style::InjectableStyleKind, utils::common::hash_f64};

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Hash, Eq)]
pub struct InjectableStyleBase {
  pub rtl: Option<String>,
  pub ltr: String,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Hash, Eq)]
pub struct InjectableStyleConstBase {
  pub rtl: Option<String>,
  pub ltr: String,
  pub const_key: String,
  pub const_value: String,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub struct InjectableStyle {
  pub ltr: String,
  pub rtl: Option<String>,
  pub priority: Option<f64>,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub(crate) struct InjectableConstStyle {
  pub ltr: String,
  pub rtl: Option<String>,
  pub priority: Option<f64>,
  pub const_key: String,
  pub const_value: String,
}

impl Hash for InjectableStyle {
  fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
    self.ltr.hash(state);
    self.rtl.hash(state);
    hash_f64(self.priority.unwrap_or(0.0));
  }
}

impl From<InjectableStyle> for InjectableStyleBase {
  fn from(style: InjectableStyle) -> Self {
    // Assuming InjectableStyleBase and InjectableStyle have similar fields
    InjectableStyleBase {
      ltr: style.ltr,
      rtl: style.rtl,
    }
  }
}

impl InjectableStyle {
  /// Creates a new InjectableStyle wrapped in Rc<InjectableStyleKind> with only LTR content.
  ///
  /// # Example
  /// ```ignore
  /// let style = InjectableStyle::regular(css_string, Some(0.5));
  /// ```
  #[inline]
  pub(crate) fn regular(ltr: String, priority: Option<f64>) -> Rc<InjectableStyleKind> {
    Rc::new(InjectableStyleKind::Regular(InjectableStyle {
      ltr,
      rtl: None,
      priority,
    }))
  }

  /// Creates a new InjectableStyle wrapped in Rc<InjectableStyleKind> with both LTR and RTL content.
  ///
  /// # Example
  /// ```ignore
  /// let style = InjectableStyle::with_rtl(ltr_css, rtl_css, Some(0.5));
  /// ```
  #[inline]
  pub(crate) fn with_rtl(ltr: String, rtl: String, priority: Option<f64>) -> Rc<InjectableStyleKind> {
    Rc::new(InjectableStyleKind::Regular(InjectableStyle {
      ltr,
      rtl: Some(rtl),
      priority,
    }))
  }
}

impl Default for InjectableStyle {
  fn default() -> Self {
    InjectableStyle {
      ltr: "".to_string(),
      rtl: None,
      priority: Some(0.0),
    }
  }
}

impl From<InjectableConstStyle> for InjectableStyleConstBase {
  fn from(style: InjectableConstStyle) -> Self {
    InjectableStyleConstBase {
      ltr: style.ltr,
      rtl: style.rtl,
      const_key: style.const_key,
      const_value: style.const_value,
    }
  }
}
impl Default for InjectableConstStyle {
  fn default() -> Self {
    InjectableConstStyle {
      ltr: "".to_string(),
      rtl: None,
      priority: Some(0.0),
      const_key: "".to_string(),
      const_value: "".to_string(),
    }
  }
}
