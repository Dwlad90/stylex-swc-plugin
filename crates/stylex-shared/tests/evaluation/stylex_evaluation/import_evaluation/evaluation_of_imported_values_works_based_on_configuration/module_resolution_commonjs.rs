use std::env;

use insta::assert_snapshot;
use stylex_shared::shared::structures::stylex_options::{StyleXOptions, StyleXOptionsParams};
use stylex_shared::{StyleXTransform, shared::structures::plugin_pass::PluginPass};
use swc_core::ecma::{
  parser::{Syntax, TsSyntax},
  transforms::testing::test,
};

use crate::utils::transform::stringify_js;

fn tranform(input: &str) -> String {
  stringify_js(
    input,
    Syntax::Typescript(TsSyntax {
      tsx: true,
      ..Default::default()
    }),
    |tr| {
      let cwd_path = std::env::current_dir().unwrap();

      let fixture_path = cwd_path.join("tests/fixture/consts");

      let mut config = StyleXOptionsParams {
        runtime_injection: Some(true),
        treeshake_compensation: Some(true),
        unstable_module_resolution: Some(StyleXOptions::get_common_js_module_resolution(Some(
          fixture_path.to_string_lossy().to_string(),
        ))),
        ..Default::default()
      };

      StyleXTransform::new_test_force_runtime_injection_with_pass(
        tr.comments.clone(),
        PluginPass {
          cwd: Some(fixture_path.clone()),
          filename: fixture_path.clone().join("input.stylex.js").into(),
        },
        Some(&mut config),
      )
    },
  )
}

#[test]
fn recognizes_ts_stylex_imports_when_resolving_js_relative_imports() {
  let input = r#"import stylex from 'stylex';
        import { MyTheme } from './input.stylex';
        const styles = stylex.create({
          red: {
            color: MyTheme.__varGroupHash__,
          }
        });
        stylex(styles.red);"#;

  let transformation = tranform(input);

  assert_snapshot!(transformation);
}
