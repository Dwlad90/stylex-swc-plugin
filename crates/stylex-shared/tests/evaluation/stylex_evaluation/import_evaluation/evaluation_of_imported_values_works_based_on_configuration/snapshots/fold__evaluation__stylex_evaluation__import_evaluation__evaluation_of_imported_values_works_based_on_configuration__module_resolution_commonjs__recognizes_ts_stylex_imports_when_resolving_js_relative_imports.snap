---
source: crates/stylex-shared/tests/evaluation/stylex_evaluation/import_evaluation/evaluation_of_imported_values_works_based_on_configuration/module_resolution_commonjs.rs
expression: transformation
---
import _inject from "@stylexjs/stylex/lib/stylex-inject";
var _inject2 = _inject;
import "./input.stylex";
import stylex from 'stylex';
import { MyTheme } from './input.stylex';
_inject2(".x2yizs4{color:x1tseezp}", 3000);
"x2yizs4";
